### ISSUE TITLE
[Feature Request] Allow users to choose ippool via CNI_ARGS

### ISSUE THREAD

**Issue by jschwinger23** [projectcalico/cni-plugin#1055]
## Expected Behavior
We are using CNI aside from K8s, and we'd like to empower users to choose the specified  ippool.
In Docker CNM system, this can be easily done by `docker run --net [ippool]` as long as we have`docker network create`d a network in line with calico ippool.

## Current Behavior
Current Calico CNI offers 3 ways to achieve the similar tasks:
1. choose specify ip address:  `CNI_ARGS` env already supports this (refer to https://github.com/containernetworking/cni/blob/master/CONVENTIONS.md#cni_args)
2. choose speicfy ippool: K8s users are able to claim the ippool via annotation (refer to https://github.com/projectcalico/cni-plugin/issues/213)
3. pre-claim available ippools: calico cni configure allows for this requirements in `ipam` block (refer to https://docs.projectcalico.org/reference/cni-plugin/configuration#ipam)

but none of the aforementioned take into account the requirements of choosing specifying ippool without k8s.

## Possible Solution
As far as I can see, we can make good use of `CNI_ARGS` env again, and it appears really sensible since we already have ability to specify ip address through that.

Therefore, I suggest that a env `CNI_ARGS=IPPOOL=[poolname]` be recognized by Calico CNI and be used for ipam.

**Comment by mgleung** [comment]
Hi @jschwinger23 , thanks for bringing this up. I know we currently have support for specifying ippools as a part of the CNI config file (https://docs.projectcalico.org/reference/cni-plugin/configuration) which I think should give you the functionality you need. Does this work or is there another use case that works better with adding it to the env variables?

**Comment by caseydavenport** [comment]
For what it is worth, CNI_ARGS is deprecated in favor of the "args" map within the CNI config json.

That said, I think Calico already supports setting IP pool on a per-network basis via the CNI config json, as you found here: https://docs.projectcalico.org/reference/cni-plugin/configuration#ipam

What do you need that the CNI config json can't already give you?

**Comment by jschwinger23** [comment]
Thanks for your reply @mgleung @caseydavenport.
The issue with CNI config json is it's supposed to be "static". Imagine that we have to modify the config on every host every time a new calico ippool is created, and I won't see it as a pleasant experience.

On the other hand, even if there won't be such a frequent change of ippool, for instance there are 3 pools all set up well in the CNI config as `"ipv4_pools": ["pool-1", "pool-2", "pool-3"]`, how can the users (in the view of orchestrator) specify what pool he/she would like to use?

If user wanna use `pool-3` only, how do the CNI-invoker pass this information to CNI? Maybe revise the CNI config in line with the user request before giving config to CNI?

Put it in another way:

In docker CNM system, users are able to specify the pool by `docker run --netowrk [pool]`;
In K8s orchestrator, users are able to specify the pool by annotation `cni.calico/ipv4pools: [pool]`;
hypothetically, should an another orchestrator integrate into calico CNI, and expose an user-level API equivalent to `docker run --net [pool]` or K8s annotation `cni.calico/ipv4pools`, how would this orchestrator tell the calico CNI which pool to use, according to the user request, rather than the static preset CNI config?

**Comment by caseydavenport** [comment]
> The issue with CNI config json is it's supposed to be "static".

This isn't actually true, and in fact the container runtime is free to mutate and manipulate the CNI config json as needed on a per-request basis. This is done, for example, by Mesos in order to pass label information to the CNI plugin on a per-container basis.

> how would this orchestrator tell the calico CNI which pool to use,

The main problem I see is that an IP pool is a Calico-specific concept, and not applicable to all CNI plugins. I don't think we need an enhancement to Calico, rather we need an enhancement to whatever container runtime to pass CIDR information via the CNI config json on a per-request basis. 

Which runtime/orchestrator are you using?

**Comment by jschwinger23** [comment]
> runtime is free to mutate and manipulate the CNI config json

Ok, that's reasonable.

> Which runtime/orchestrator are you using?

I'm working on integrating dockerd into CNI.

Change the subject completely, I was wondering why calico CNI applies a(n) (arguably) double standard for K8s and non-K8s?

When K8s wanted to allow users to specify ippool in annotation, calico CNI surprisingly calls the K8s API to fetch the annotation, in a way  too specific and causing bidirectional dependency: k8s(cri) calls calico CNI, then calico CNI calls k8s.
https://github.com/projectcalico/cni-plugin/blob/f73028f8c4b6327826b3d2cd5351706c903786d3/pkg/k8s/k8s.go#L191

And when it comes to another runtime/orchestrator, let's say, how about letting calico CNI fetch docker containers' labels to choose which ippool to use, I'm sure the proposal will be rejected in favor of "mutate and manipulate the CNI config json".

I'm not saying "mutate and manipulate the CNI config json" is a bad idea, actually it makes sense very much. Just curious why K8s guys have such a great privilege to make integration by changing calico CNI?

**Comment by caseydavenport** [comment]
> When K8s wanted to allow users to specify ippool in annotation, calico CNI surprisingly calls the K8s API to fetch the annotation

K8s didn't want to add this, the Calico team did. And so, we did, in the way that was available to us. There's nothing surprising about this, and there's no double-standard either. When we went to implement this, we went through a design process and considered a number of options.

If you would like to extend a feature to dockerd, then we need to go through that same design process, which is what I am trying to do with you on this issue - not debate whether or not we implemented Kubernetes-specific features "fairly". Our options are the following:

1. Have the container runtime (docker) pass IP pools via the CNI config json using the existing field.
2. Have container runtime use CNI_ARGS to do this
3. If there is an API available, have the CNI plugin call back in to the docker API to get the information it needs.

(1) is my preferred option. (2) is deprecated and so that's a no-go. (3) mirrors what the plugin does for Kubernetes. 

I don't want (3) because it requires additional configuration and knowledge to be passed to the CNI plugin and is thus harder to implement and in general, I'd like to start _removing_ orchestrator-specific code rather than introducing more. 

> Just curious why K8s guys have such a great privilege to make integration by changing calico CNI?

If you'd like to submit a PR to Kubernetes / containerd / crio / podman / dockershim / etc to pass per-pod annotations through the  CNI args map, I would be delighted to removed the code that calls into the Kubernetes API. But I don't see that happening any time soon.
